src/lang-promql/parser/parser.ts

```diff
@@ -0,0 +1,298 @@
+    }
+    this.expectType(expr, NodeTypeVector, 'aggregation expression');
+    // get the parameter of the aggregation operator
+    const params = walkThrough(node, FunctionCallBody, FunctionCallArgs, FunctionCallArgs, Expr);
+    if (aggregateOp.type.id === Topk || aggregateOp.type.id === Bottomk || aggregateOp.type.id === Quantile) {
+      if (!params) {
+        this.addDiagnostic(node, 'no parameter found');
+        return;
+      }
+      this.expectType(params, NodeTypeScalar, 'aggregation parameter');
+    }
+    if (aggregateOp.type.id === CountValues) {
+      if (!params) {
+        this.addDiagnostic(node, 'no parameter found');
+        return;
+      }
+      this.expectType(params, NodeTypeString, 'aggregation parameter');
+    }
+  }
+
+  private checkBinaryExpr(node: Subtree): void {
+    // Following the definition of the BinaryExpr, the left and the right
+    // expression are respectively the first and last child
+    // https://github.com/promlabs/lezer-promql/blob/master/src/promql.grammar#L52
+    const lExpr = node.firstChild;
+    const rExpr = node.lastChild;
+    if (!lExpr || !rExpr) {
+      this.addDiagnostic(node, 'left or right expression is missing in binary expression');
+      return;
+    }
+    const lt = this.checkAST(lExpr);
+    const rt = this.checkAST(rExpr);
+    const boolModifierUsed = walkThrough(node, BinModifier, GroupModifiers, BoolModifier, Bool);
+    const isComparisonOperator = childExist(node, Eql, Neq, Lte, Lss, Gte, Gtr);
+    const isSetOperator = childExist(node, And, Or, Unless);
+
+    // BOOL modifier check
+    if (boolModifierUsed) {
+      if (!isComparisonOperator) {
+        this.addDiagnostic(node, 'bool modifier can only be used on comparison operators');
+      }
+      if (lt !== NodeTypeScalar || rt !== NodeTypeScalar) {
+        this.addDiagnostic(node, 'comparisons between other things than scalar cannot use BOOL modifier');
+      }
+    } else {
+      if (lt === NodeTypeScalar && rt === NodeTypeScalar) {
+        this.addDiagnostic(node, 'comparisons between scalars must use BOOL modifier');
+      }
+    }
+    // TODO missing check regarding cardManyToOne or cardOneToMany
+    // TODO missing check regarding the matching label
+    if (lt !== NodeTypeScalar && lt !== NodeTypeVector) {
+      this.addDiagnostic(lExpr, 'binary expression must contain only scalar and instant vector types');
+    }
+    if (lt !== NodeTypeScalar && lt !== NodeTypeVector) {
+      this.addDiagnostic(rExpr, 'binary expression must contain only scalar and instant vector types');
+    }
+    if ((lt === NodeTypeScalar || rt === NodeTypeScalar) && isSetOperator) {
+      this.addDiagnostic(node, 'set operator not allowed in binary scalar expression');
+    }
+  }
+
+  private expectType(node: Subtree, want: NodeType, context: string): void {
+    const t = this.checkAST(node);
+    if (t !== want) {
+      this.addDiagnostic(node, 'expected type %s in %s, got %s', want, context, t);
```

Comment:
^^ that's because I didn't know about it actually. I just learned this way a couple hours today ( I was searching how to print the complete tree) and I didn't think it could replace the method `printf()`

Reply:
I'll replace it